CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc

```diff
@@ -4,50 +4,102 @@
       if (w > bestweight){
-          bestweight=w;
-          iVertex=index;
-        }        
+        bestweight=w;
+        iVertex=index;
+      }        
       index++;
   }
+  
+  bool useTime = useTiming_;
+  if (edm::isNotFinite(time) || timeReso<1e-6) {
+    useTime = false;
+    time = 0.;
+    timeReso = -1.;
+  }
+  
+  if (preferHighRanked_) {
+    for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
+        int ivtx = iv - vertices.begin();
+        if (iVertex == ivtx) return std::pair<int,PrimaryVertexAssignment::Quality>(ivtx,PrimaryVertexAssignment::UsedInFit);
+        
+        double dz = std::abs(track->dz(iv->position()));
+        double dt = std::abs(time-iv->t());
+        
+        bool useTimeVtx = useTime && iv->tError()>0.;
+        
+        if ((dz < maxDzForPrimaryAssignment_ or dz/track->dzError() < maxDzSigForPrimaryAssignment_ ) and (!useTimeVtx or dt/timeReso < maxDtSigForPrimaryAssignment_)) {
+          return std::pair<int,PrimaryVertexAssignment::Quality>(ivtx,PrimaryVertexAssignment::PrimaryDz);
+        }               
+    }
+  }
 
-
+    
   if(iVertex >= 0 ) return std::pair<int,PrimaryVertexAssignment::Quality>(iVertex,PrimaryVertexAssignment::UsedInFit);
-
-    double dzmin = 1e99;
-    int vtxIdMinDz = -1;
-    for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
-      double dz = std::abs(track->dz(iv->position()));
-      if(dz<dzmin) {
-        dzmin = dz;
-        vtxIdMinDz = iv-vertices.begin();
-      }
-   }
+    
+  double distmin = std::numeric_limits<double>::max();
+  double dzmin = std::numeric_limits<double>::max();
+  double dtmin = std::numeric_limits<double>::max();
+  int vtxIdMinDist = -1;
+  for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
+    double dz = std::abs(track->dz(iv->position()));
+    double dt = std::abs(time-iv->t());
+    
+    double dzsig = dz/track->dzError();
+    double dist = dzsig*dzsig;
+    
+    bool useTimeVtx = useTime && iv->tError()>0.;
+    if (useTime && !useTimeVtx) {
+      dt = timeReso;
+    }
+    
+    if (useTime) {
+      double dtsig = dt/timeReso;
+              
+      dist += dtsig*dtsig;
+    }
+    if(dist<distmin) {
+      distmin = dist;
+      dzmin = dz;
+      dtmin = dt;
+      vtxIdMinDist = iv-vertices.begin();
+    }
+  }
+      
   // first use "closest in Z" with tight cuts (targetting primary particles)
-    float dzE=sqrt(track->dzError()*track->dzError()+vertices[vtxIdMinDz].covariance(2,2));
-    if(vtxIdMinDz>=0 and (dzmin < maxDzForPrimaryAssignment_ and dzmin/dzE < maxDzSigForPrimaryAssignment_  and track->dzError()<maxDzErrorForPrimaryAssignment_))
+    const float add_cov = vtxIdMinDist >= 0 ? vertices[vtxIdMinDist].covariance(2,2) : 0.f;
+    const float dzE=sqrt(track->dzError()*track->dzError()+add_cov);
+    if(vtxIdMinDist>=0 and 
+       (dzmin < maxDzForPrimaryAssignment_ and dzmin/dzE < maxDzSigForPrimaryAssignment_  and track->dzError()<maxDzErrorForPrimaryAssignment_) and
```

Comment:
There is in fact a small issue which has been introduced here when forward porting this code to 93x and resolving merge conflicts with 9d480b860ffa309bd6a83ccc5408bbd842448c3d

It would not affect the non-timing vertices, but could have a small impact on the sorting of the 4d vertices.  I'll commit a fix momentarily.

Reply:
Sorry, that's not true, I was misreading/misremembering what this code does.

Everything which is here is fine, with indeed a small inconsistency between the handling of z and t, since the vertex error is used here for z, but not for t.  I would prefer to leave it like this for now, also because we need to study more closely the pulls on the time errors of the 4d vertices.